src/main/scala/it/pagopa/interop/authorizationmanagement/api/impl/KeyApiServiceImpl.scala

```diff
@@ -5,74 +5,71 @@ import akka.actor.typed.ActorSystem
+import scala.concurrent.{Await, ExecutionContext, Future}
 
 final case class KeyApiServiceImpl(
   system: ActorSystem[_],
   sharding: ClusterSharding,
-  entity: Entity[Command, ShardingEnvelope[Command]]
-) extends KeyApiService
+  entity: Entity[Command, ShardingEnvelope[Command]],
+  dateTimeSupplier: OffsetDateTimeSupplier
+)(implicit ec: ExecutionContext)
+    extends KeyApiService
     with Validation {
 
-  private val logger = Logger.takingImplicit[ContextFieldsToLog](this.getClass)
+  private implicit val logger: LoggerTakingImplicit[ContextFieldsToLog] =
+    Logger.takingImplicit[ContextFieldsToLog](this.getClass)
 
-  private val settings: ClusterShardingSettings = shardingSettings(entity, system)
+  private implicit val settings: ClusterShardingSettings = shardingSettings(entity, system)
+  private implicit val implicitSharding: ClusterSharding = sharding
 
-  /** Code: 201, Message: Keys created
-    * Code: 400, Message: Missing Required Information
-    * Code: 404, Message: Client id not found, DataType: Problem
-    */
   override def createKeys(clientId: String, key: Seq[KeySeed])(implicit
     toEntityMarshallerKeysResponse: ToEntityMarshaller[KeysResponse],
     toEntityMarshallerProblem: ToEntityMarshaller[Problem],
     contexts: Seq[(String, String)]
   ): Route = authorize(ADMIN_ROLE, SECURITY_ROLE) {
-    // TODO consider a preauthorize for user rights validations...
-    logger.info("Creating keys for client {}", clientId)
+    val operationLabel: String = s"Creating keys for client $clientId"
+    logger.info(operationLabel)
+
     val validatedPayload: ValidatedNel[String, Seq[ValidKey]] = validateKeys(key)
 
-    validatedPayload.andThen(k => validateWithCurrentKeys(k, keysIdentifiers)) match {
-      case Valid(validKeys) =>
-        val commander: EntityRef[Command]             =
-          sharding.entityRefFor(KeyPersistentBehavior.TypeKey, getShard(clientId, settings.numberOfShards))
-        val result: Future[StatusReply[KeysResponse]] =
-          commander.ask(ref => AddKeys(clientId, validKeys, ref))
-        onSuccess(result) {
-          case statusReply if statusReply.isSuccess => createKeys201(statusReply.getValue)
-          case statusReply                          =>
-            logger.error(s"Error while creating keys for client $clientId", statusReply.getError)
-            createKeys400(problemOf(StatusCodes.BadRequest, CreateKeysBadRequest(clientId: String)))
-        }
-
-      case Invalid(errors) =>
-        val errorsStr = errors.toList.mkString(", ")
-        logger.error(s"Error while creating keys for client $clientId - $errorsStr")
-        createKeys400(problemOf(StatusCodes.BadRequest, CreateKeysInvalid(clientId)))
-    }
+    val result: Future[KeysResponse] = for {
+      validKeys      <- validatedPayload
+        .andThen(k =>
+          validateWithCurrentKeys(k, keysIdentifiers) // TODO Not sure this validateWithCurrentKeys should be done
```

Comment:
the `validateWithCurrentKeys` verifies that the given kid has not been used in ANY other client (not necessarily owned by the requester).
To me this is a very strong assumption (and maybe a remote possible security issue).
Opinions?

Reply:
Could you explain which issues you are talking about? Preventing duplicating a key in the entire set sounds like a security feature.

Reply:
after sleeping on it I changed my mind, it should not be an issue